Add support for the Web Extensions Permissions API.

kiaraarose opened the pull request:
#### 290f26cb9bdcb5fe32a70d108af0eb6d4702720f
<pre>
Add support for the Web Extensions Permissions API.
<a href="https://bugs.webkit.org/show_bug.cgi?id=248555">https://bugs.webkit.org/show_bug.cgi?id=248555</a>

Reviewed by Timothy Hatcher.

This patch adds the implementation for the Permissions APIs defined in the  WebExtensionAPIPermissions idl file.
This includes consulting the WebExtensionContext for information on the current permissions the
extension has access to, and reporting that information back to WebProcess.

* Source/WTF/wtf/HashSet.h:
(WTF::W&gt;::isSubset):
Checks if the current set is a subset of a given set.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm: Added.
(WebKit::WebExtensionContext::getAll):
Get all the current granted permissions and match patterns for the extension.

(WebKit::WebExtensionContext::contains):
Checks that the current permissions/matches match the given permissions and match patterns.

(WebKit::WebExtensionContext::permissionsRequest):
<a href="https://webkit.org/b/250135">https://webkit.org/b/250135</a> is being used to track adding support for permissions.request().

(WebKit::WebExtensionContext::permissionsRemove):
Removes access for specified permissions and origins.

(WebKit::WebExtensionContext::parseMatchPatterns):
Parsing the input HashSet of origins and populates a MatchPatternSet.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::hasPermissions):
Checks if the extension has access to a set of permissions and match patterns.

(WebKit::WebExtensionContext::scheduleBackgroundContentToUnload):
Removed code that caused an infinite loop. Implementation being tracked in <a href="https://webkit.org/b/246483.">https://webkit.org/b/246483.</a>

* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::isPropertyAllowed):
(WebKit::WebExtensionAPINamespace::permissions):
Add support for the permissions API.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm: Added.
(WebKit::WebExtensionAPIPermissions::getAll):
(WebKit::WebExtensionAPIPermissions::contains):
(WebKit::WebExtensionAPIPermissions::request):
(WebKit::WebExtensionAPIPermissions::remove):
(WebKit::WebExtensionAPIPermissions::parseDetailsDictionary):
Populates a set of permissions and origins to pass to the UIProcess.

(WebKit::WebExtensionAPIPermissions::verifyRequestedPermissions):
Checks that the permissions being requested/removed are valid.

(WebKit::WebExtensionAPIPermissions::validatePermissionsDetails):
Validates the data passed in. Check that the permissions requested are supported.

(WebKit::WebExtensionAPIPermissions::onAdded):
(WebKit::WebExtensionAPIPermissions::onRemoved):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIPermissions.h:
Copied from Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIPermissions.idl:

Canonical link: <a href="https://commits.webkit.org/258552@main">https://commits.webkit.org/258552@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/080200ba16e3d268ed7f60f6a4d8c2d31163f167

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102330 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11468 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35396 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111639 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/171812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12468 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2387 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94659 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109349 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108111 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9538 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/92805 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/37282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24290 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/79029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/92638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4981 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25713 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/88895 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/2645 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5119 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2155 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29546 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11149 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45206 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/91819 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6872 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20544 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3102 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
```diff
@@ -230,6 +231,9 @@ class WebExtensionContext : public API::ObjectImpl<API::Object::Type::WebExtensi
     void sendToProcessesForEvent(WebExtensionEventListenerType, const T& message);
 
+    template<typename T>
+    bool isSubset(HashSet<T> setA, HashSet<T> setB);
```
Is there a method that I'm not aware of that already does this? If not, should I add this to HashSet.h?

xeenon replied:
There is not, and I've wanted it before. You should put this in `HashSet.h` as `bool isSubset(HashSet<T> other)`.

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -0,0 +1,124 @@
+{
+    RELEASE_LOG(Extensions, "permissions.request()");
+
+    // FIXME: Check to see if the request is being made from a user gesture.
```
Add link to FIXME. <https://webkit.org/b/250135

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -0,0 +1,196 @@
+
+void WebExtensionContext::getAll(WebPageProxyIdentifier identifier, CompletionHandler<void(Vector<String>, Vector<String>)>&& completionHandler)
+{
+    Vector<String> permissions, origins;
```
Check to see if the web proxy identifier matches the background web view.

xeenon replied:
Why do you need to check for the background web view?

kiaraarose replied:
Removed.

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -0,0 +1,196 @@
+void WebExtensionContext::contains(WebPageProxyIdentifier identifier, HashSet<String> permissions, HashSet<String> origins, CompletionHandler<void(bool)>&& completionHandler)
+{
+    if (m_backgroundWebView.get()._page->identifier() != identifier) {
+        completionHandler(false);
```
Should there be an error message passed back here?

xeenon replied:
I'm not sure why this check is needed.

kiaraarose replied:
I wasn't sure if it was needed or not. I'll remove them.

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -0,0 +1,196 @@
+        return;
+    }
+
+    if (!verifyRequestedPermissions(permissions, origins, matchPatterns, @"remove()", &errorMessage)) {
```
Currently, in Safari we don't report back an error message and just fail silently to match Chrome's behavior. Although, I think passing back an error message in this case might be helpful to developers.

xeenon replied:
Safari does report the error, it is doing it on the web process side via `validatePermissionDetails()`, not in the UI process. It needs to be in the web process to synchronously throw the error.

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1246,17 +1269,6 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
 
     // FIXME: <https://webkit.org/b/246483> Don't unload the background page if there are open ports, pending website requests, or if an inspector window is open.
-
-    if (m_backgroundWebView.get()._isBeingInspected) {
```
This code was causing an infinite loop so I removed it. Should be revisited in https://webkit.org/b/246483.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -0,0 +1,124 @@
+    HashSet<String> permissions, origins;
+    parseDetailsDictionary(details, permissions, origins);
+
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::Contains(page->webPageProxyIdentifier(), permissions, origins), [ protectedThis = Ref { *this }, callback = WTFMove(callback) ](bool containsPermissions) {
```
Drop the spaces inside the brackets: `[protectedThis = Ref { *this }, callback = WTFMove(callback)]`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -0,0 +1,124 @@
+    }, extensionContext().identifier().toUInt64());
+}
+
+void WebExtensionAPIPermissions::contains(WebPage* page, NSDictionary *details, Ref<WebExtensionCallbackHandler>&& callback, NSString **)
```
The `NSString **` error string should be used. The existing code's `validatePermissionDetails()` does this. You can file a bug and add a FIXME for this.

kiaraarose replied:
Must've missed that! The updated PR will have this change.

xeenon commented on Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIPermissions.idl:
```diff
@@ -26,24 +26,23 @@
 
     // Checks if the extension has the specified permissions.
-    void getAll([Optional, CallbackHandler] function callback);
+    [NeedsPage] void getAll([Optional, CallbackHandler] function callback);
```
Shouldn't need the page for these.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -0,0 +1,124 @@
+{
+    RELEASE_LOG(Extensions, "permissions.getAll()");
+
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::GetAll(page->webPageProxyIdentifier()), [ protectedThis = Ref { *this }, callback = WTFMove(callback) ](Vector<String> permissions, Vector<String> origins) {
```
Drop the spaces inside the brackets: `[protectedThis = Ref { *this }, callback = WTFMove(callback)]`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -0,0 +1,124 @@
+    HashSet<String> permissions, origins;
+    parseDetailsDictionary(details, permissions, origins);
+
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::Remove(page->webPageProxyIdentifier(), permissions, origins), [ protectedThis = Ref { *this }, callback = WTFMove(callback) ](bool success, String errorMessage) {
```
Drop the spaces inside the brackets: `[protectedThis = Ref { *this }, callback = WTFMove(callback)]`

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
```diff
@@ -291,6 +295,14 @@ class WebExtensionContext : public API::ObjectImpl<API::Object::Type::WebExtensi
     void removeListener(WebPageProxyIdentifier, WebExtensionEventListenerType);
 
+    // Permissions APIs
+    void getAll(WebPageProxyIdentifier, CompletionHandler<void(Vector<String> permissions, Vector<String> origins)>&&);
```
We will run into namespace clashes with the methods names at some point. We should prefix the messages and the methods with the namespace, so `permissionsGetAll()`. Like I did with `test` above.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -0,0 +1,196 @@
+    MatchPatternSet patterns;
+    for (auto& origin : origins) {
+        auto pattern = WebExtensionMatchPattern::getOrCreate(static_cast<NSString *>(origin));
+        if (!pattern || !pattern->isValid()) {
```
You can just check for `null`. `getOrCreate()` already does an `isValid()` check and returns `nullptr`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -731,6 +731,29 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    for (auto& pattern : matchPatterns) {
+        bool matchFound = false;
+        for (auto& grantedPattern : currentPermissionMatchPatterns()) {
+            if (pattern->matchesPattern(grantedPattern)) {
```
This needs to pass the `{ WebExtensionMatchPattern::Options::IgnorePaths }` option to `matchesPattern()`.

kiaraarose committed:
Add support for the Web Extensions Permissions API.
https://bugs.webkit.org/show_bug.cgi?id=248555

Reviewed by Timothy Hatcher.

This patch adds the implementation for the Permissions APIs defined in the  WebExtensionAPIPermissions idl file.
This includes consulting the WebExtensionContext for information on the current permissions the
extension has access to, and reporting that information back to WebProcess.

* Source/WTF/wtf/HashSet.h:
(WTF::W>::isSubset):
Checks if the current set is a subset of a given set.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm: Added.
(WebKit::WebExtensionContext::getAll):
Get all the current granted permissions and match patterns for the extension.

(WebKit::WebExtensionContext::contains):
Checks that the current permissions/matches match the given permissions and match patterns.

(WebKit::WebExtensionContext::permissionsRequest):
https://webkit.org/b/250135 is being used to track adding support for permissions.request().

(WebKit::WebExtensionContext::permissionsRemove):
Removes access for specified permissions and origins.

(WebKit::WebExtensionContext::parseMatchPatterns):
Parsing the input HashSet of origins and populates a MatchPatternSet.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::hasPermissions):
Checks if the extension has access to a set of permissions and match patterns.

(WebKit::WebExtensionContext::scheduleBackgroundContentToUnload):
Removed code that caused an infinite loop. Implementation being tracked in https://webkit.org/b/246483.

* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::isPropertyAllowed):
(WebKit::WebExtensionAPINamespace::permissions):
Add support for the permissions API.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm: Added.
(WebKit::WebExtensionAPIPermissions::getAll):
(WebKit::WebExtensionAPIPermissions::contains):
(WebKit::WebExtensionAPIPermissions::request):
(WebKit::WebExtensionAPIPermissions::remove):
(WebKit::WebExtensionAPIPermissions::parseDetailsDictionary):
Populates a set of permissions and origins to pass to the UIProcess.

(WebKit::WebExtensionAPIPermissions::verifyRequestedPermissions):
Checks that the permissions being requested/removed are valid.

(WebKit::WebExtensionAPIPermissions::validatePermissionsDetails):
Validates the data passed in. Check that the permissions requested are supported.

(WebKit::WebExtensionAPIPermissions::onAdded):
(WebKit::WebExtensionAPIPermissions::onRemoved):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIPermissions.h:
Copied from Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIPermissions.idl:

Canonical link: https://commits.webkit.org/258552@main
